src/Tmds.Ssh/Curve25519KeyExchange.cs

```diff
@@ -0,0 +1,215 @@
+
+        // Receive ECDH_REPLY.
+        using Packet ecdhReplyMsg = await context.ReceivePacketAsync(MessageId.SSH_MSG_KEX_ECDH_REPLY, firstPacket.Move(), ct).ConfigureAwait(false);
+        var ecdhReply = ParceEcdhReply(ecdhReplyMsg);
+
+        // Verify received key is valid.
+        connectionInfo.ServerKey = new HostKey(ecdhReply.public_host_key);
+        await hostKeyVerification.VerifyAsync(connectionInfo, ct).ConfigureAwait(false);
+
+        var publicHostKey = PublicKey.CreateFromSshKey(ecdhReply.public_host_key);
+        if (publicHostKey is RsaPublicKey rsaPublicKey && rsaPublicKey.KeySize < input.MinimumRSAKeySize)
+        {
+            throw new ConnectFailedException(ConnectFailedReason.KeyExchangeFailed, $"Server RSA key size {rsaPublicKey.KeySize} is less than {input.MinimumRSAKeySize}.", connectionInfo);
+        }
+
+        // Compute shared secret.
+        BigInteger sharedSecret;
+        try
+        {
+            sharedSecret = DeriveSharedSecret(x25519KeyPair.Private, new X25519PublicKeyParameters(ecdhReply.q_s));
+        }
+        catch (Exception ex)
+        {
+            throw new ConnectFailedException(ConnectFailedReason.KeyExchangeFailed, "Cannot determine shared secret.", connectionInfo, ex);
+        }
+
+        // Generate exchange hash.
+        byte[] exchangeHash = CalculateExchangeHash(sequencePool, input.ConnectionInfo, input.ClientKexInitMsg, input.ServerKexInitMsg, ecdhReply.public_host_key.Data, q_c, ecdhReply.q_s, sharedSecret);
+
+        // Verify the server's signature.
+        if (!publicHostKey.VerifySignature(input.HostKeyAlgorithms, exchangeHash, ecdhReply.exchange_hash_signature))
+        {
+            throw new ConnectFailedException(ConnectFailedReason.KeyExchangeFailed, "Signature does not match host key.", connectionInfo);
+        }
+
+        byte[] sessionId = input.ConnectionInfo.SessionId ?? exchangeHash;
+        byte[] initialIVC2S = CalculateKey(sequencePool, sharedSecret, exchangeHash, (byte)'A', sessionId, input.InitialIVC2SLength);
+        byte[] initialIVS2C = CalculateKey(sequencePool, sharedSecret, exchangeHash, (byte)'B', sessionId, input.InitialIVS2CLength);
+        byte[] encryptionKeyC2S = CalculateKey(sequencePool, sharedSecret, exchangeHash, (byte)'C', sessionId, input.EncryptionKeyC2SLength);
+        byte[] encryptionKeyS2C = CalculateKey(sequencePool, sharedSecret, exchangeHash, (byte)'D', sessionId, input.EncryptionKeyS2CLength);
+        byte[] integrityKeyC2S = CalculateKey(sequencePool, sharedSecret, exchangeHash, (byte)'E', sessionId, input.IntegrityKeyC2SLength);
+        byte[] integrityKeyS2C = CalculateKey(sequencePool, sharedSecret, exchangeHash, (byte)'F', sessionId, input.IntegrityKeyS2CLength);
+
+        return new KeyExchangeOutput(exchangeHash,
+            initialIVS2C, encryptionKeyS2C, integrityKeyS2C,
+            initialIVC2S, encryptionKeyC2S, integrityKeyC2S);
```

Comment:
```
return new CalculateKeyExchangeOutput(...);
```
using `CalculateKeyExchangeOutput` from base class.

Reply:
Note: the `sharedSecret` is `mpint` when ECDH and Curve25519, but is octet string when hybrid. See https://www.ietf.org/archive/id/draft-ietf-sshm-ntruprime-ssh-01.html

Reply:
We can change `CalculateKey` to take the shared secret as a `ReadOnlySpan<byte>`.

In `BigIntegerExtensions` we can add something like:

```cs
    public static byte[] ToMPIntString(this BigInteger integer)
        => integer == BigInteger.Zero ? Array.Empty<byte>() : integer.ToByteArray(isUnsigned: false, isBigEndian: true);
```

and call it here to change the `BigInteger` into an octet string.

Reply:
`ToMPIntString` -> `ToMPIntByteArray` is a better name. (string in the method name may be confusing)

Reply:
> and call it here to change the `BigInteger` into an octet string.

Sorry I'm not quite understand. Regarding hybrid kex, the derived secret is octet string. Where does `BigInteger` come from?

Reply:
If we'd like to use `CalculateKeyExchangeOutput` in the hybrid kex, we want to pass it an octet string. So we can change the implementation (in `ECDHKeyExchange` and `Curve25519KeyExchange`) to convert the `BigInteger` shared secret into a byte array (the octet string content) that gets passed to `CalculateKeyExchangeOutput`.

Reply:
The change for this is:
```diff
diff --git a/src/Tmds.Ssh/BigIntegerExtensions.cs b/src/Tmds.Ssh/BigIntegerExtensions.cs
index 98124f2..5c189a3 100644
--- a/src/Tmds.Ssh/BigIntegerExtensions.cs
+++ b/src/Tmds.Ssh/BigIntegerExtensions.cs
@@ -19,4 +19,7 @@ public static byte[] ToBEByteArray(this BigInteger integer, bool isUnsigned, int
         }
         return array;
     }
+
+    public static byte[] ToMPIntByteArray(this BigInteger integer)
+        => integer == BigInteger.Zero ? Array.Empty<byte>() : integer.ToByteArray(isUnsigned: false, isBigEndian: true);
 }
\ No newline at end of file
diff --git a/src/Tmds.Ssh/Curve25519KeyExchange.cs b/src/Tmds.Ssh/Curve25519KeyExchange.cs
index 4b74ce1..b60a4d2 100644
--- a/src/Tmds.Ssh/Curve25519KeyExchange.cs
+++ b/src/Tmds.Ssh/Curve25519KeyExchange.cs
@@ -44,7 +44,7 @@ public override async Task<KeyExchangeOutput> TryExchangeAsync(KeyExchangeContex
         var publicHostKey = await VerifyHostKeyAsync(hostKeyVerification, input, ecdhReply.public_host_key, ct).ConfigureAwait(false);
 
         // Compute shared secret.
-        BigInteger sharedSecret;
+        byte[] sharedSecret;
         try
         {
             sharedSecret = DeriveSharedSecret(x25519KeyPair.Private, new X25519PublicKeyParameters(ecdhReply.q_s));
@@ -63,7 +63,7 @@ public override async Task<KeyExchangeOutput> TryExchangeAsync(KeyExchangeContex
         return CalculateKeyExchangeOutput(input, sequencePool, sharedSecret, exchangeHash, _hashAlgorithmName);
     }
 
-    private static byte[] CalculateExchangeHash(SequencePool sequencePool, SshConnectionInfo connectionInfo, ReadOnlyPacket clientKexInitMsg, ReadOnlyPacket serverKexInitMsg, byte[] public_host_key, byte[] q_c, byte[] q_s, BigInteger sharedSecret, HashAlgorithmName hashAlgorithmName)
+    private static byte[] CalculateExchangeHash(SequencePool sequencePool, SshConnectionInfo connectionInfo, ReadOnlyPacket clientKexInitMsg, ReadOnlyPacket serverKexInitMsg, byte[] public_host_key, byte[] q_c, byte[] q_s, byte[] sharedSecret, HashAlgorithmName hashAlgorithmName)
     {
         /*
             string   V_C, client's identification string (CR and LF excluded)
@@ -84,7 +84,7 @@ private static byte[] CalculateExchangeHash(SequencePool sequencePool, SshConnec
         writer.WriteString(public_host_key);
         writer.WriteString(q_c);
         writer.WriteString(q_s);
-        writer.WriteMPInt(sharedSecret);
+        writer.WriteString(sharedSecret);
 
         using IncrementalHash hash = IncrementalHash.CreateHash(hashAlgorithmName);
         foreach (var segment in sequence.AsReadOnlySequence())
@@ -94,7 +94,7 @@ private static byte[] CalculateExchangeHash(SequencePool sequencePool, SshConnec
         return hash.GetHashAndReset();
     }
 
-    private static BigInteger DeriveSharedSecret(AsymmetricKeyParameter privateKey, AsymmetricKeyParameter peerPublicKey)
+    private static byte[] DeriveSharedSecret(AsymmetricKeyParameter privateKey, AsymmetricKeyParameter peerPublicKey)
     {
         var keyAgreement = new X25519Agreement();
         keyAgreement.Init(privateKey);
@@ -103,7 +103,7 @@ private static BigInteger DeriveSharedSecret(AsymmetricKeyParameter privateKey,
         keyAgreement.CalculateAgreement(peerPublicKey, rawSecretAgreement);
         var sharedSecret = rawSecretAgreement.ToBigInteger();
         rawSecretAgreement.AsSpan().Clear();
-        return sharedSecret;
+        return sharedSecret.ToMPIntByteArray();
     }
 
     private static Packet CreateEcdhInitMessage(SequencePool sequencePool, ReadOnlySpan<byte> q_c)
diff --git a/src/Tmds.Ssh/ECDHKeyExchange.cs b/src/Tmds.Ssh/ECDHKeyExchange.cs
index 21d3913..47bf2b9 100644
--- a/src/Tmds.Ssh/ECDHKeyExchange.cs
+++ b/src/Tmds.Ssh/ECDHKeyExchange.cs
@@ -39,7 +39,7 @@ public override async Task<KeyExchangeOutput> TryExchangeAsync(KeyExchangeContex
         var publicHostKey = await VerifyHostKeyAsync(hostKeyVerification, input, ecdhReply.public_host_key, ct).ConfigureAwait(false);
 
         // Compute shared secret.
-        BigInteger sharedSecret;
+        byte[] sharedSecret;
         try
         {
             sharedSecret = DeriveSharedSecret(ecdh, _ecCurve, ecdhReply.q_s);
@@ -58,7 +58,7 @@ public override async Task<KeyExchangeOutput> TryExchangeAsync(KeyExchangeContex
         return CalculateKeyExchangeOutput(input, sequencePool, sharedSecret, exchangeHash, _hashAlgorithmName);
     }
 
-    private static byte[] CalculateExchangeHash(SequencePool sequencePool, SshConnectionInfo connectionInfo, ReadOnlyPacket clientKexInitMsg, ReadOnlyPacket serverKexInitMsg, byte[] public_host_key, ECPoint q_c, ECPoint q_s, BigInteger sharedSecret, HashAlgorithmName hashAlgorithmName)
+    private static byte[] CalculateExchangeHash(SequencePool sequencePool, SshConnectionInfo connectionInfo, ReadOnlyPacket clientKexInitMsg, ReadOnlyPacket serverKexInitMsg, byte[] public_host_key, ECPoint q_c, ECPoint q_s, byte[] sharedSecret, HashAlgorithmName hashAlgorithmName)
     {
         /*
             string   V_C, client's identification string (CR and LF excluded)
@@ -79,7 +79,7 @@ private static byte[] CalculateExchangeHash(SequencePool sequencePool, SshConnec
         writer.WriteString(public_host_key);
         writer.WriteString(q_c);
         writer.WriteString(q_s);
-        writer.WriteMPInt(sharedSecret);
+        writer.WriteString(sharedSecret);
 
         using IncrementalHash hash = IncrementalHash.CreateHash(hashAlgorithmName);
         foreach (var segment in sequence.AsReadOnlySequence())
@@ -89,7 +89,7 @@ private static byte[] CalculateExchangeHash(SequencePool sequencePool, SshConnec
         return hash.GetHashAndReset();
     }
 
-    private static BigInteger DeriveSharedSecret(ECDiffieHellman ecdh, ECCurve curve, ECPoint q)
+    private static byte[] DeriveSharedSecret(ECDiffieHellman ecdh, ECCurve curve, ECPoint q)
     {
         ECParameters parameters = new ECParameters
         {
@@ -101,7 +101,7 @@ private static BigInteger DeriveSharedSecret(ECDiffieHellman ecdh, ECCurve curve
         byte[] rawSecretAgreement = ecdh.DeriveRawSecretAgreement(peerPublicKey);
         var sharedSecret = rawSecretAgreement.ToBigInteger();
         rawSecretAgreement.AsSpan().Clear();
-        return sharedSecret;
+        return sharedSecret.ToMPIntByteArray();
     }
 
     private static Packet CreateEcdhInitMessage(SequencePool sequencePool, ECPoint q_c)
diff --git a/src/Tmds.Ssh/KeyExchange.cs b/src/Tmds.Ssh/KeyExchange.cs
index 25408b8..9ceab22 100644
--- a/src/Tmds.Ssh/KeyExchange.cs
+++ b/src/Tmds.Ssh/KeyExchange.cs
@@ -32,7 +32,7 @@ protected static void VerifySignature(PublicKey publicHostKey, IReadOnlyList<Nam
         }
     }
 
-    protected static KeyExchangeOutput CalculateKeyExchangeOutput(KeyExchangeInput input, SequencePool sequencePool, BigInteger sharedSecret, byte[] exchangeHash, HashAlgorithmName hashAlgorithmName)
+    protected static KeyExchangeOutput CalculateKeyExchangeOutput(KeyExchangeInput input, SequencePool sequencePool, byte[] sharedSecret, byte[] exchangeHash, HashAlgorithmName hashAlgorithmName)
     {
         byte[] sessionId = input.ConnectionInfo.SessionId ?? exchangeHash;
         byte[] initialIVC2S = CalculateKey(sequencePool, sharedSecret, exchangeHash, (byte)'A', sessionId, input.InitialIVC2SLength, hashAlgorithmName);
@@ -47,7 +47,7 @@ protected static KeyExchangeOutput CalculateKeyExchangeOutput(KeyExchangeInput i
             initialIVC2S, encryptionKeyC2S, integrityKeyC2S);
     }
 
-    protected static byte[] CalculateKey(SequencePool sequencePool, BigInteger sharedSecret, byte[] exchangeHash, byte c, byte[] sessionId, int keyLength, HashAlgorithmName hashAlgorithmName)
+    protected static byte[] CalculateKey(SequencePool sequencePool, byte[] sharedSecret, byte[] exchangeHash, byte c, byte[] sessionId, int keyLength, HashAlgorithmName hashAlgorithmName)
     {
         // https://tools.ietf.org/html/rfc4253#section-7.2
 
@@ -57,7 +57,7 @@ protected static byte[] CalculateKey(SequencePool sequencePool, BigInteger share
         // HASH(K || H || c || session_id)
         using Sequence sequence = sequencePool.RentSequence();
         var writer = new SequenceWriter(sequence);
-        writer.WriteMPInt(sharedSecret);
+        writer.WriteString(sharedSecret);
         writer.Write(exchangeHash);
         writer.WriteByte(c);
         writer.Write(sessionId);
@@ -76,7 +76,7 @@ protected static byte[] CalculateKey(SequencePool sequencePool, BigInteger share
 
             // K3 = HASH(K || H || K1 || K2)
             writer = new SequenceWriter(sequence);
-            writer.WriteMPInt(sharedSecret);
+            writer.WriteString(sharedSecret);
             writer.Write(exchangeHash);
             writer.Write(key.AsSpan(0, keyOffset));

```

Reply:
Do you want me to apply the change in this PR or the upcoming hybrid kex PR?

Reply:
Let's include it here, so what is provided from the base class is more "generic".

Reply:
Done